### PR TITLE
rpc: Return block hash & height in getbalances, gettransaction and getwalletinfo

### LINKED ISSUE #18567: Return block hash with wallet calls
Some wallet calls (like getbalances) need to lock the chainstate to make sure they are caught up on the latest block (which may have been reported to the users in getblockchaininfo or similar).

Those calls could benefit if the block hash they were valid on was included in the response. For example, immature balance can only change when the chainstate changes. (Assuming no keys are imported)

It seems those calls can include the block hash they are caught up with for free in the response. They could simply include `m_last_block_processed` in the response.

I haven't checked back with other developers to see if they like this approach, but to start simple only a single call could be updated for now. E.g. `getbalances`.

#### Useful skills:

Basic understanding of threads, the Bitcoin Core wallet and the Bitcoin Core RPC interface.

#### Want to work on this issue?

The purpose of the `good first issue` label is to highlight which issues are suitable for a new contributor without a deep understanding of the codebase.

You do not need to request permission to start working on this. You are encouraged to comment on the issue if you are planning to work on it. This will help other contributors monitor which issues are actively being addressed and is also an effective way to request assistance if and when you need it.

For guidance on contributing, please read [CONTRIBUTING.md](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md) before opening your pull request.

### PR DESCRIPTION
**Pull request by aureleoules**
Reopens #18570 and closes #18567.
I have rebased the original PR.
Not sure why the original got closed as it was about to get merged.

### CODE REVIEW

**Comment by luke-jr** [comment]
As I said previously,

> I'm not sure this is the right approach, interface-wise, especially for calls like gettransaction, where the info gets embedded in the JSON transaction itself.
>
>Maybe it'd be better to just guarantee atomicity for batch requests?

File: test/functional/wallet_balance.py [hunk 1]
```diff
@@ -296,5 +303,27 @@ def test_balances(*, fee_node_1=0):
...
+        # Tests the lastprocessedblock JSON object in getbalances, getwalletinfo
+        # and gettransaction by checking for valid hex strings and by comparing
+        # the hashes & heights between generated blocks.
+        self.log.info("Test getbalances returns expected lastprocessedblock json object")
+        prev_hash = self.nodes[0].getbestblockhash()
+        prev_height = self.nodes[0].getblock(prev_hash)['height']
+        self.generatetoaddress(self.nodes[0], 5, self.nodes[0].get_deterministic_priv_key().address)
+        lastblock = self.nodes[0].getbalances()['lastprocessedblock']
+        assert_is_hash_string(lastblock['hash'])
+        assert_equal((prev_hash == lastblock['hash']), False)
```
Comment (jarolrod) [inline test/functional/wallet_balance.py:315]: there's also `assert_not_equal` that could be used

File: test/functional/wallet_balance.py [hunk 2]
```diff
@@ -296,5 +303,27 @@ def test_balances(*, fee_node_1=0):
...
+        assert_greater_than(lastblock['height'], prev_height)
+
+        self.log.info("Test getwalletinfo returns expected lastprocessedblock json object")
+        walletinfo = self.nodes[0].getwalletinfo()
+        assert_greater_than(walletinfo['lastprocessedblock']['height'], 0)
+        assert_is_hash_string(walletinfo['lastprocessedblock']['hash'])
+
+        self.log.info("Test gettransaction returns expected lastprocessedblock json object")
```
Comment (jarolrod) [inline test/functional/wallet_balance.py:323]: should also have a test on the height to ensure that this is in fact the "expected" `lastprocessedblock` json object?

**Review by jarolrod — COMMENTED** [review]
Concept ACK

This does what it says it does. I think it's fine to add these fields, I can imagine it would be useful. Tested and sanity checked the code, and ran tests locally.

File: test/functional/wallet_balance.py (hunk 1 above)
Comment (aureleoules) [reply test/functional/wallet_balance.py:315]: `assert_not_equal` does not exist :neutral_face:

Comment (jarolrod) [reply test/functional/wallet_balance.py:315]: right, sorry, mind was mixed up while reviewing: https://github.com/bitcoin/bitcoin/pull/23127

File: test/functional/wallet_balance.py (hunk 2 above)
Comment (aureleoules) [reply test/functional/wallet_balance.py:325]: Done.

**Review by willcl-ark — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/commit/37a05b4f6bd768873cdfaaf0b417ef7d29c60f97

Returns the blockheight and hash the aforementioned RPCs were run against. Feels like a win for users of the RPCs.

**Comment by achow101** [comment]
ACK 37a05b4f6bd768873cdfaaf0b417ef7d29c60f97

> Maybe it'd be better to just guarantee atomicity for batch requests?

I don't really see how that would be possible to implement in a way that wouldn't be detrimental or fragile. It seems like it would have to be implemented with an explicit list of RPCs that must be done atomically in order to avoid any RPC just locking things up. Such a list could fall out of date as new RPCs are added and people forget, so it would be kind of fragile.

**Comment by achow101** [comment]
ACK 710b83938ab5bbc4bd324d8b2e69461a2a1d2eec

**Pull request merged**